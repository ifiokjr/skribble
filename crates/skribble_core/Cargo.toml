[package]
name = "skribble_core"
version = { workspace = true }
authors = { workspace = true }
categories = { workspace = true }
documentation = "https://docs.rs/skribble_core"
edition = { workspace = true }
homepage = { workspace = true }
include = ["src/**/*.rs", "Cargo.toml", "readme.md"]
keywords = ["build", "atomic", "css"]
license = { workspace = true }
readme = "readme.md"
repository = { workspace = true }
rust-version = { workspace = true }
description = "Core features for the skribble atomic css library"

[dependencies]
derivative = { workspace = true }
derive_more = { workspace = true }
doc-comment = { workspace = true }
globset = { workspace = true }
heck = { workspace = true }
indent_write = { workspace = true }
indexmap = { features = ["serde"], workspace = true }
lazy_static = { workspace = true }
lightningcss = { workspace = true }
readonly = { workspace = true }
regex = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
skribble_color = { workspace = true, features = ["serde"] }
thiserror = { workspace = true }
typed-builder = { workspace = true }
vfs = { workspace = true }

[dev-dependencies]
insta = { features = ["json"], workspace = true }
