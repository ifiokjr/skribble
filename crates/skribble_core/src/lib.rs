#![deny(clippy::all)]
#![forbid(clippy::indexing_slicing)]

pub use config::*;
pub use constants::*;
pub use error::*;
pub use plugin::*;
pub use runner::*;
pub use utils::*;

mod config;
mod constants;
mod error;
mod plugin;
mod runner;
mod utils;
