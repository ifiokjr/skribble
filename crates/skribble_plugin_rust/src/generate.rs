use indexmap::IndexMap;
use indexmap::IndexSet;

use super::*;

fn generate_media_queries(
  config: &MergedConfig,

  method_names: &mut IndexSet<String>,
  sections: &mut Vec<String>,
  struct_names_map: &mut IndexMap<String, usize>,
  trait_names: &mut Vec<String>,
) {
  for (key, map) in config.media_queries.iter() {
    let mut section = Vec::<String>::new();
    let trait_name = format!("MediaQuery{}", key.to_pascal_case());
    let struct_name = format!("{trait_name}Child");
    section.push(generate_struct(&struct_name));
    section.push(generate_impl_skribble_value(&struct_name));

    let mut methods = vec![format!("pub trait {trait_name}: SkribbleValue {{")];

    for (name, media_query) in map.iter() {
      let method_name = get_method_name(name, method_names);
      let css_docs = wrap_indent(
        wrap_docs(wrap_in_code_block(
          media_query_docs(&media_query.query),
          "css",
        )),
        1,
      );

      if let Some(ref description) = media_query.description {
        methods.push(wrap_indent(wrap_docs(description), 1));
        methods.push(wrap_indent(wrap_docs("\n"), 1));
      }

      methods.push(css_docs);
      methods.push(wrap_indent(
        format!("#[inline]\nfn {method_name}(&self) -> {struct_name} {{"),
        1,
      ));
      methods.push(wrap_indent(
        format!("{struct_name}::from_ref(self.append_to_skribble_value(\"{name}\"))"),
        2,
      ));
      methods.push(wrap_indent("}", 1));
    }

    methods.push("}".into());
    section.push(methods.join("\n"));

    trait_names.push(trait_name);
    struct_names_map.insert(struct_name, trait_names.len());
    sections.push(section.join("\n"));
  }
}

fn media_query_docs(query: impl AsRef<str>) -> String {
  let query = query.as_ref();
  format!("@media {query} {{\n  /* ... */\n}}")
}

fn modifier_docs(values: &[String]) -> String {
  let value = values.join(", ");
  format!("{value} {{\n  /* ... */\n}}")
}

fn generate_modifiers(
  config: &MergedConfig,
  method_names: &mut IndexSet<String>,
  sections: &mut Vec<String>,
  struct_names_map: &mut IndexMap<String, usize>,
  trait_names: &mut Vec<String>,
) {
  for (key, map) in config.modifiers.iter() {
    let mut section = Vec::<String>::new();
    let trait_name = format!("Modifier{}", key.to_pascal_case());
    let struct_name = format!("{trait_name}Child");
    section.push(generate_struct(&struct_name));
    section.push(generate_impl_skribble_value(&struct_name));

    let mut methods = vec![format!("pub trait {trait_name}: SkribbleValue {{")];

    for (name, modifier) in map.iter() {
      let method_name = get_method_name(name, method_names);
      let css_docs = wrap_indent(
        wrap_docs(wrap_in_code_block(modifier_docs(&modifier.values), "css")),
        1,
      );

      if let Some(ref description) = modifier.description {
        methods.push(wrap_indent(wrap_docs(description), 1));
        methods.push(wrap_indent(wrap_docs("\n"), 1));
      }

      methods.push(css_docs);
      methods.push(wrap_indent(
        format!("#[inline]\nfn {method_name}(&self) -> {struct_name} {{"),
        1,
      ));
      methods.push(wrap_indent(
        format!("{struct_name}::from_ref(self.append_to_skribble_value(\"{name}\"))"),
        2,
      ));
      methods.push(wrap_indent("}", 1));
    }

    methods.push("}".into());
    section.push(methods.join("\n"));

    trait_names.push(trait_name);
    struct_names_map.insert(struct_name, trait_names.len());
    sections.push(section.join("\n"));
  }
}

fn generate_keyframes(config: &MergedConfig, sections: &mut Vec<String>, name: impl AsRef<str>) {
  let name = name.as_ref();

  sections.push(format!("pub trait {name}: SkribbleValue {{"));

  for (name, keyframe) in config.keyframes.iter() {
    let method_name = safe_method_name(name);
    let css_docs = wrap_indent(
      wrap_docs(wrap_in_code_block(keyframe_docs(keyframe, config), "css")),
      1,
    );

    if let Some(ref description) = keyframe.description {
      sections.push(wrap_indent(wrap_docs(description), 1));
      sections.push(wrap_indent(wrap_docs("\n"), 1));
    }

    sections.push(css_docs);

    sections.push(wrap_indent(
      format!("#[inline]\nfn {method_name}(&self) -> String {{"),
      1,
    ));

    sections.push(wrap_indent(
      format!("self.append_string_to_skribble_value(\"{name}\")"),
      2,
    ));

    sections.push(wrap_indent("}", 1));
  }

  sections.push("}".into());
}

fn keyframe_docs(keyframe: &Keyframe, config: &MergedConfig) -> String {
  let mut content = Vec::<String>::new();

  content.push(format!("@keyframes {} {{", keyframe.name));

  for (key, map) in keyframe.rules.iter() {
    content.push(wrap_indent(format!("{key} {{"), 1));

    for (name, value) in map.iter() {
      let css = wrap_indent(
        Placeholder::normalize(format!("{}: {};", name, value), config),
        2,
      );
      content.push(css)
    }

    content.push(wrap_indent("}", 1));
  }

  content.push("}".into());

  content.join("\n")
}

fn generate_value_sets(config: &MergedConfig, sections: &mut Vec<String>) {
  for (name, value_set) in config.value_sets.iter() {
    let value_set_trait_name = get_value_set_trait_name(name);
    sections.push(format!(
      "pub trait {value_set_trait_name}: SkribbleValue {{"
    ));

    for (value_name, _) in value_set.values.iter() {
      let method_name = safe_method_name(value_name);

      if let Some(ref description) = value_set.description {
        sections.push(wrap_indent(wrap_docs(description), 1));
      }

      sections.push(wrap_indent(
        format!("#[inline]\nfn {method_name}(&self) -> String {{"),
        1,
      ));

      sections.push(wrap_indent(
        format!("self.append_string_to_skribble_value(\"{value_name}\")"),
        2,
      ));

      sections.push(wrap_indent("}", 1));
    }

    sections.push("}".into());
  }
}

fn generate_named_classes(
  config: &MergedConfig,

  sections: &mut Vec<String>,
  trait_names: &mut Vec<String>,
) {
  sections.push("pub trait NamedClasses: SkribbleValue {".into());

  for (class_name, class) in config.classes.iter() {
    let method_name = safe_method_name(class_name);

    if let Some(ref description) = class.description {
      sections.push(wrap_indent(wrap_docs(description), 1));
    }

    sections.push(wrap_indent(
      format!("#[inline]\nfn {method_name}(&self) -> String {{"),
      1,
    ));

    sections.push(wrap_indent(
      format!("self.append_string_to_skribble_value(\"{class_name}\")"),
      2,
    ));

    sections.push(wrap_indent("}", 1));
  }

  trait_names.push("NamedClasses".into());
  sections.push("}".into());
}

const ATOM_TRAIT_NAME: &str = "Atom";

fn generate_atoms(
  config: &MergedConfig,

  method_names: &mut IndexSet<String>,
  sections: &mut Vec<String>,
  trait_names: &mut Vec<String>,
) {
  let mut struct_content = Vec::<String>::new();
  let mut trait_content = Vec::<String>::new();

  // parent modifiers
  trait_content.push(format!("pub trait {ATOM_TRAIT_NAME}: SkribbleValue {{"));

  for (name, modifier) in config.atoms.iter() {
    let method_name = get_method_name(name, method_names);
    let atom_struct_name = format!("Atom{}", name.to_pascal_case());

    struct_content.push(generate_struct(&atom_struct_name));
    struct_content.push(generate_impl_skribble_value(&atom_struct_name));

    match modifier.values {
      LinkedValues::Color(ref color) => {
        struct_content.push(format!("impl Color for {atom_struct_name} {{}}"));

        if !color.ignore_palette {
          struct_content.push(format!("impl Palette for {atom_struct_name} {{}}"));
        }
      }
      LinkedValues::Values(ref value_set) => {
        for value_set_name in value_set.iter() {
          let value_set_trait_name = get_value_set_trait_name(&value_set_name.value);

          struct_content.push(format!(
            "impl {value_set_trait_name} for {atom_struct_name} {{}}",
          ));
        }
      }
      LinkedValues::Keyframes => {
        let keyframe_trait_name = get_keyframe_trait_name(&atom_struct_name);
        generate_keyframes(config, sections, &keyframe_trait_name);

        struct_content.push(format!(
          "impl {keyframe_trait_name} for {atom_struct_name} {{}}",
        ));
      }
    }

    if let Some(ref description) = modifier.description {
      trait_content.push(wrap_indent(wrap_docs(description), 1));
    }

    trait_content.push(wrap_indent(
      format!("#[inline]\nfn {method_name}(&self) -> {atom_struct_name} {{"),
      1,
    ));

    trait_content.push(wrap_indent(
      format!("{atom_struct_name}::from_ref(self.append_to_skribble_value(\"{name}\"))"),
      2,
    ));

    trait_content.push(wrap_indent("}", 1));
  }

  trait_content.push("}".into());

  trait_names.push(ATOM_TRAIT_NAME.into());
  sections.push(struct_content.join("\n"));
  sections.push(trait_content.join("\n"));
}

fn generate_palette(config: &MergedConfig, sections: &mut Vec<String>) {
  sections.push("pub trait Palette: SkribbleValue {".into());

  for (name, _) in config.palette.iter() {
    let method_name = safe_method_name(name);

    sections.push(wrap_indent(
      format!("#[inline]\nfn {method_name}(&self) -> String {{"),
      1,
    ));

    sections.push(wrap_indent(
      format!("self.append_string_to_skribble_value(\"{name}\")"),
      2,
    ));

    sections.push(wrap_indent("}", 1));
  }

  sections.push("}".into());
}

fn generate_css_variables(
  config: &MergedConfig,
  variable_prefix: impl AsRef<str>,
  sections: &mut Vec<String>,
) {
  let mut entries = vec![
    indoc!(
      "
    pub fn vars() -> CssVariables {
      CssVariables
    }
    pub struct CssVariables;
    impl CssVariables {"
    )
    .into(),
  ];
  let mut colors = vec!["pub trait Color: SkribbleValue {".into()];

  for (name, css_variable) in config.css_variables.iter() {
    let method_name = safe_method_name(name);
    let variable_name = css_variable.get_variable(variable_prefix.as_ref());
    let css_docs = wrap_indent(
      wrap_docs(wrap_in_code_block(
        css_property_docs(
          &variable_name,
          &css_variable.syntax,
          &css_variable.value,
          config,
        ),
        "css",
      )),
      1,
    );

    if let Some(ref description) = css_variable.description {
      entries.push(wrap_indent(wrap_docs(description), 1));
      entries.push(wrap_indent(wrap_docs("\n"), 1));
    }

    entries.push(css_docs.clone());

    entries.push(wrap_indent(
      format!("#[inline]\npub fn {method_name}(&self) -> String {{"),
      1,
    ));

    entries.push(wrap_indent(format!("\"{variable_name}\".into()",), 2));

    entries.push(wrap_indent("}", 1));

    if css_variable.syntax.is_color() {
      if let Some(ref description) = css_variable.description {
        colors.push(wrap_indent(wrap_docs(description), 1));
        colors.push(wrap_indent(wrap_docs("\n"), 1));
      }

      colors.push(css_docs);

      colors.push(wrap_indent(
        format!("#[inline]\nfn {method_name}(&self) -> String {{"),
        1,
      ));
      colors.push(wrap_indent(
        format!("self.append_string_to_skribble_value(\"{name}\")"),
        2,
      ));
      colors.push(wrap_indent("}", 1))
    }
  }

  entries.push("}".into());
  colors.push("}".into());
  sections.push(entries.join("\n"));
  sections.push(colors.join("\n"));
}

fn wrap_indent(content: impl AsRef<str>, level: u8) -> String {
  let mut result = content.as_ref().to_string();
  let indent_style = IndentStyle::default();

  for _ in 1..=level {
    result = indent(result, indent_style);
  }

  result
}

fn wrap_docs(content: impl AsRef<str>) -> String {
  let mut result = vec![];
  for line in content.as_ref().lines() {
    result.push(format!("/// {line}"));
  }

  result.join("\n")
}

fn css_property_docs(
  variable_name: impl AsRef<str>,
  syntax: &PropertySyntax,
  initial_value: &Option<String>,
  config: &MergedConfig,
) -> String {
  let variable_name = variable_name.as_ref();
  let default_initial_value = "/* */".into();
  let initial_value = Placeholder::normalize(
    initial_value.as_ref().unwrap_or(&default_initial_value),
    config,
  );
  format!(
    "@property {variable_name} {{\n  syntax: \"{syntax}\";\n  inherits: false;\n  initial-value: \
     {initial_value};\n}}"
  )
}

fn wrap_in_code_block(content: impl AsRef<str>, r#type: impl AsRef<str>) -> String {
  format!("```{}\n{}\n```", r#type.as_ref(), content.as_ref(),)
}

fn get_value_set_trait_name(value_set_name: impl Into<String>) -> String {
  format!("ValueSet{}", value_set_name.into().to_pascal_case())
}

fn get_keyframe_trait_name(atom_name: impl Into<String>) -> String {
  format!("KeyframeSet{}", atom_name.into().to_pascal_case())
}

fn generate_impl_skribble_value(name: impl AsRef<str>) -> String {
  format!(
    indoc!(
      "
      impl SkribbleValue for {} {{
        #[inline]
        fn from_ref(value: impl AsRef<str>) -> Self {{
          Self(value.as_ref().to_string())
        }}
        #[inline]
        fn get_skribble_value(&self) -> &String {{
          &self.0
        }}
      }}"
    ),
    name.as_ref()
  )
}

fn generate_struct_implementations(
  struct_names_map: &IndexMap<String, usize>,
  trait_names: &[String],
  sections: &mut Vec<String>,
) {
  let mut content = Vec::<String>::new();
  for (struct_name, min_index) in struct_names_map.iter() {
    for (index, trait_name) in trait_names.iter().enumerate() {
      if *min_index <= index {
        content.push(format!("impl {trait_name} for {struct_name} {{}}",));
      }
    }
  }

  sections.push(content.join("\n"));
}

fn generate_struct(name: impl AsRef<str>) -> String {
  let name = name.as_ref();
  format!("pub struct {name}(String);")
}

fn get_method_name(name: impl AsRef<str>, method_names: &mut IndexSet<String>) -> String {
  let method_name = safe_method_name(name);
  let mut index = 0;
  let mut current_method_name = method_name.clone();
  loop {
    if method_names.contains(&current_method_name) {
      index += 1;
      current_method_name = format!("{}{}", method_name, index);
      continue;
    }

    method_names.insert(current_method_name.clone());
    break;
  }

  current_method_name
}

fn safe_method_name(name: impl AsRef<str>) -> String {
  let name = name.as_ref();

  let prefix = match name.chars().next() {
    Some(first_char) if first_char.is_ascii_digit() => "n_",
    Some(first_char) if !first_char.is_ascii_alphabetic() => {
      match first_char {
        '-' => "minus_",
        '+' => "plus_",
        _ => "__",
      }
    }
    _ => "",
  };

  let method_name = format!("{prefix}{}", name.to_snake_case());

  if RESERVED_WORDS.contains(&method_name.as_str()) {
    return format!("r#{}", method_name);
  }

  if method_name.is_empty() {
    return "__".into();
  }

  method_name
}

const RESERVED_WORDS: &[&str] = &[
  "abstract", "alignof", "as", "become", "box", "break", "const", "continue", "crate", "do",
  "else", "enum", "extern", "false", "final", "fn", "for", "if", "impl", "in", "let", "loop",
  "macro", "match", "mod", "move", "mut", "offsetof", "override", "priv", "proc", "pub", "pure",
  "ref", "return", "Self", "self", "sizeof", "static", "struct", "super", "trait", "true", "type",
  "typeof", "unsafe", "unsized", "use", "virtual", "where", "while", "yield",
];

const HEADER: &str = r#"#![allow(unused)]
// This file was generated by skribble.
use private::SkribbleValue;
pub fn sk() -> SkribbleRoot {
  SkribbleRoot::from_ref("")
}
pub struct SkribbleRoot(String);
impl SkribbleValue for SkribbleRoot {
  #[inline]
  fn from_ref(value: impl AsRef<str>) -> Self {
    Self(value.as_ref().to_string())
  }
  #[inline]
  fn get_skribble_value(&self) -> &String {
    &self.0
  }
}
mod private {
  #[doc(hidden)]
  pub trait SkribbleValue {
    fn from_ref(value: impl AsRef<str>) -> Self;
    fn get_skribble_value(&self) -> &String;
    #[inline]
    fn append_to_skribble_value(&self, value: impl AsRef<str>) -> String {
      let current_value = self.get_skribble_value();
      let prefix = if current_value.is_empty() {
        "".into()
      } else {
        format!("{current_value}:")
      };

      format!("{}{}", prefix, value.as_ref())
    }
    #[inline]
    fn append_string_to_skribble_value(&self, value: impl AsRef<str>) -> String {
      format!("{}:${}", self.get_skribble_value(), value.as_ref())
    }
  }
}"#;

fn combine_sections_with_header(sections: Vec<String>) -> String {
  // let indent_style = IndentStyle::default();
  // let mut content = vec![];
  // content.push("pub use generated_skribble_module::sk;".into());
  // content.push("pub use generated_skribble_module::vars;".into());
  // content.push("use generated_skribble_module::*;".into());
  // content.push("mod generated_skribble_module {".into());
  // content.push(indent(HEADER, indent_style));
  // content.push(indent(sections.join("\n"), indent_style));
  // content.push("}".into());

  // content.join("\n")
  format!("{HEADER}\n{}", sections.join("\n"))
}

pub(crate) fn generate_file_contents(config: &MergedConfig) -> String {
  let mut method_names: IndexSet<String> = indexset! {};
  let mut sections = Vec::<String>::new();
  let mut trait_names = vec![];
  let mut struct_names_map: IndexMap<String, usize> = indexmap! { "SkribbleRoot".into() => 0 };

  generate_css_variables(config, &config.options().variable_prefix, &mut sections);

  // media queries
  generate_media_queries(
    config,
    &mut method_names,
    &mut sections,
    &mut struct_names_map,
    &mut trait_names,
  );

  generate_modifiers(
    config,
    &mut method_names,
    &mut sections,
    &mut struct_names_map,
    &mut trait_names,
  );

  generate_value_sets(config, &mut sections);
  generate_palette(config, &mut sections);

  generate_atoms(config, &mut method_names, &mut sections, &mut trait_names);

  generate_named_classes(config, &mut sections, &mut trait_names);

  // Add the implementation for each of the structs.
  generate_struct_implementations(&struct_names_map, &trait_names, &mut sections);
  combine_sections_with_header(sections)
}
