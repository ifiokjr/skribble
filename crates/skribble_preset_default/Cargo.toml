[package]
name = "skribble_preset_default"
version = { workspace = true }
authors = { workspace = true }
documentation = "https://docs.rs/skribble_preset_default"
edition = { workspace = true }
homepage = { workspace = true }
include = ["src/**/*.rs", "Cargo.toml", "readme.md"]
keywords = ["build", "atomic", "css", "preset", "skribble"]
license = { workspace = true }
readme = "readme.md"
repository = { workspace = true }
rust-version = { workspace = true }
description = "This plugin provides a default preset for Skribble which is similar to `tailwind`, `windi` and `unocss`."

[dependencies]
heck = { workspace = true }
indexmap = { workspace = true }
lazy_static = { workspace = true }
serde = { workspace = true }
skribble_core = { workspace = true }
typed-builder = { workspace = true }

[dev-dependencies]
insta = { workspace = true }
