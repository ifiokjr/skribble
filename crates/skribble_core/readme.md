# skribble_core

> core library for skribble

<br />

[![Crate][crate-image]][crate-link] [![Docs][docs-image]][docs-link] [![Status][ci-status-image]][ci-status-link] [![Unlicense][unlicense-image]][unlicense-link]

## Installation

```toml
[dependencies]
skribble_core = "0.0.0"
```

### Usage

```rust
use skribble_core::*;
```

[crate-image]: https://img.shields.io/crates/v/skribble_core.svg
[crate-link]: https://crates.io/crates/skribble_core
[docs-image]: https://docs.rs/skribble_core/badge.svg
[docs-link]: https://docs.rs/skribble_core
[ci-status-image]: https://github.com/ifiokjr/skribble/workflows/ci/badge.svg
[ci-status-link]: https://github.com/ifiokjr/skribble/actions?query=workflow:ci
[unlicense-image]: https://img.shields.io/badge/license-Unlicence-blue.svg
[unlicense-link]: https://opensource.org/license/unlicense
