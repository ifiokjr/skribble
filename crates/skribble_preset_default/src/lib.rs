#![deny(clippy::all)]

use data::*;
use enums::*;
use serde::Deserialize;
use serde::Serialize;
use skribble_core::*;
use typed_builder::TypedBuilder;

pub(crate) mod data;
mod enums;

#[derive(Debug, Clone, Default, Deserialize, TypedBuilder, Serialize)]
pub struct PresetDefault {
  /// Choose the palette colors from either Tailwind or OpenColors.
  #[builder(default, setter(into))]
  pub palette: PaletteType,
  /// Choose how `light` and `dark` mode are handled. via `prefers-color-scheme`
  /// or `class` based.
  #[builder(default, setter(into))]
  pub dark_mode: DarkMode,
}

impl Plugin for PresetDefault {
  fn get_id(&self) -> String {
    "skribble_preset_default".into()
  }

  fn mutate_config(&mut self, config: &mut ConfigEnum) -> AnyResult {
    match config {
      ConfigEnum::Palette(ref mut palette) => {
        self.update_palette(palette);
      }
      ConfigEnum::MediaQueries(ref mut media_queries) => {
        self.update_media_queries(media_queries);
      }
      ConfigEnum::ParentModifiers(ref mut parent_modifiers) => {
        self.update_parent_modifiers(parent_modifiers);
      }
      ConfigEnum::Modifiers(ref mut modifiers) => {
        self.update_modifiers(modifiers);
      }
      ConfigEnum::CssVariables(ref mut css_variables) => {
        self.update_css_variables(css_variables);
      }
      _ => {}
    }

    Ok(())
  }

  fn get_name(&self) -> String {
    self.get_id()
  }

  fn get_description(&self) -> String {
    "".into()
  }
}

impl PresetDefault {
  fn update_media_queries(&self, media_queries: &mut MediaQueries) {
    media_queries.extend(MEDIA_QUERIES.clone());

    if self.dark_mode == DarkMode::Media {
      media_queries.extend(DARK_MEDIA_QUERIES.clone());
    }
  }

  fn update_palette(&self, palette: &mut Palette) {
    palette.extend(self.palette.palette());
  }

  fn update_parent_modifiers(&self, parent_modifiers: &mut ParentModifiers) {
    if self.dark_mode == DarkMode::Class {
      parent_modifiers.extend(DARK_PARENT_MODIFIERS.clone());
    }

    parent_modifiers.extend(PARENT_MODIFIERS.clone());
  }

  fn update_modifiers(&self, modifiers: &mut Modifiers) {
    modifiers.extend(MODIFIERS.clone());
  }

  fn update_css_variables(&self, css_variables: &mut CssVariables) {
    css_variables.extend(COLOR_CSS_VARIABLES.clone());
    css_variables.extend(CSS_VARIABLES.clone());
  }
}
