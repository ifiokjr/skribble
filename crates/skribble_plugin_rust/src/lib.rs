#![deny(clippy::all)]
#![deny(clippy::indexing_slicing)]

use std::process::Command;
use std::process::Stdio;

use generate::generate_file_contents;
use heck::ToPascalCase;
use heck::ToSnakeCase;
use indexmap::indexmap;
use indexmap::indexset;
use indoc::indoc;
use serde::Deserialize;
use serde::Serialize;
use skribble_core::*;
use typed_builder::TypedBuilder;

mod generate;

/// This plugin generates `rust` code from the configuration.
#[derive(Debug, Clone, Default, Deserialize, TypedBuilder, Serialize)]
pub struct RustPlugin {
  /// The formatter command.
  /// e.g. `dprint`
  #[builder(default, setter(into, strip_option))]
  pub formatter: Option<String>,

  /// The formatter arguments.
  /// e.g. `["fmt", "--stdin", "file.rs"]`
  #[builder(default, setter(into))]
  pub formatter_args: Vec<String>,
}

impl Plugin for RustPlugin {
  fn get_id(&self) -> String {
    "skribble_plugin_rust".into()
  }

  fn generate_code(&self, config: &MergedConfig, options: &Options) -> AnyResult<GeneratedFiles> {
    let mut contents = generate_file_contents(config, options);

    if let Some(ref formatter) = self.formatter {
      let input = Command::new("echo")
        .arg(&contents)
        .stdout(Stdio::piped())
        .spawn()?;

      if let Some(stdout) = input.stdout {
        let output = Command::new(formatter)
          .args(&self.formatter_args)
          .stdin(stdout)
          .stdout(std::process::Stdio::piped())
          .output()?;
        contents = String::from_utf8(output.stdout)?;
      }
    }

    let mut files = GeneratedFiles::default();
    files.push(
      GeneratedFile::builder()
        .path("./src/skribble.rs")
        .content(contents)
        .build(),
    );

    Ok(files)
  }

  fn get_description(&self) -> String {
    "This plugin provides support for generating rust code from your `skribble` configuration."
      .into()
  }
}

#[cfg(test)]
mod tests {
  use skribble_core::*;
  use skribble_preset_default::PresetDefault;

  use super::*;

  #[test]
  fn default_can_be_added_to_runner() {
    let default_preset = PresetDefault::builder().build();
    let rust_plugin = RustPlugin::builder().build();

    let config: StyleConfig = StyleConfig::builder()
      .plugins(vec![
        PluginContainer::from(default_preset),
        PluginContainer::from(rust_plugin),
      ])
      .build();

    let mut runner = SkribbleRunner::new(config);
    let _ = runner.run();
    let result = runner.generate().unwrap();
    let generated = result.get(0).unwrap();
    let content = &generated.content;
    insta::assert_display_snapshot!(content);
  }
}
