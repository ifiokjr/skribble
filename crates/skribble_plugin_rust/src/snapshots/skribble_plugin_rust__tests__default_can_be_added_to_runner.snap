---
source: crates/skribble_plugin_rust/src/lib.rs
expression: content
---
// This file was generated by skribble.
pub fn sk() -> SkribbleRoot {
  SkribbleRoot::from_ref("")
}
pub struct SkribbleRoot(String);
impl SkribbleValue for SkribbleRoot {
  fn from_ref(value: impl AsRef<str>) -> Self {
    Self(value.as_ref().to_string())
  }

  fn get_skribble_value(&self) -> &String {
    &self.0
  }
}
pub trait SkribbleAtomValue {
  fn from_ref(value: impl AsRef<str>) -> Self;
  fn get_skribble_value(&self) -> &String;
  fn append_to_skribble_value(&self, value: impl AsRef<str>) -> String {
    format!("{}{}", self.get_skribble_value(), value.as_ref())
  }
}
pub trait SkribbleValue {
  fn from_ref(value: impl AsRef<str>) -> Self;
  fn get_skribble_value(&self) -> &String;
  fn append_to_skribble_value(&self, value: impl AsRef<str>) -> String {
    format!("{}{}", self.get_skribble_value(), value.as_ref())
  }
}

pub struct MediaQueryDeviceCategoriesChild(String);
impl SkribbleValue for MediaQueryDeviceCategoriesChild {
  fn from_ref(value: impl AsRef<str>) -> Self {
    Self(value.as_ref().to_string())
  }
  fn get_skribble_value(&self) -> &String {
    &self.0
  }
}

pub trait MediaQueryDeviceCategories: SkribbleValue {
  /// The media query for devices with a screen.
  fn screen(&self) -> MediaQueryDeviceCategoriesChild {
    MediaQueryDeviceCategoriesChild::from_ref(self.append_to_skribble_value("screen"))
  }
  /// The media query for devices with a printer.
  fn print(&self) -> MediaQueryDeviceCategoriesChild {
    MediaQueryDeviceCategoriesChild::from_ref(self.append_to_skribble_value("print"))
  }
}
pub struct MediaQueryBreakpointsChild(String);
impl SkribbleValue for MediaQueryBreakpointsChild {
  fn from_ref(value: impl AsRef<str>) -> Self {
    Self(value.as_ref().to_string())
  }
  fn get_skribble_value(&self) -> &String {
    &self.0
  }
}

pub trait MediaQueryBreakpoints: SkribbleValue {
  /// The breakpoint for devices with screen size greater than tiny.
  fn sm(&self) -> MediaQueryBreakpointsChild {
    MediaQueryBreakpointsChild::from_ref(self.append_to_skribble_value("sm"))
  }
  /// The breakpoint for devices screen size greater than medium
  fn md(&self) -> MediaQueryBreakpointsChild {
    MediaQueryBreakpointsChild::from_ref(self.append_to_skribble_value("md"))
  }
  /// The breakpoint for devices screen size greater than large
  fn lg(&self) -> MediaQueryBreakpointsChild {
    MediaQueryBreakpointsChild::from_ref(self.append_to_skribble_value("lg"))
  }
  /// The breakpoint for devices screen size greater than extra large
  fn xl(&self) -> MediaQueryBreakpointsChild {
    MediaQueryBreakpointsChild::from_ref(self.append_to_skribble_value("xl"))
  }
  /// The breakpoint for devices screen size greater than xxl
  fn xxl(&self) -> MediaQueryBreakpointsChild {
    MediaQueryBreakpointsChild::from_ref(self.append_to_skribble_value("xxl"))
  }
}
pub struct MediaQueryOrientationChild(String);
impl SkribbleValue for MediaQueryOrientationChild {
  fn from_ref(value: impl AsRef<str>) -> Self {
    Self(value.as_ref().to_string())
  }
  fn get_skribble_value(&self) -> &String {
    &self.0
  }
}

pub trait MediaQueryOrientation: SkribbleValue {
  /// The media query for devices with a portrait orientation.
  fn portrait(&self) -> MediaQueryOrientationChild {
    MediaQueryOrientationChild::from_ref(self.append_to_skribble_value("portrait"))
  }
  /// The media query for devices with a landscape orientation.
  fn landscape(&self) -> MediaQueryOrientationChild {
    MediaQueryOrientationChild::from_ref(self.append_to_skribble_value("landscape"))
  }
}
pub struct MediaQueryMotionChild(String);
impl SkribbleValue for MediaQueryMotionChild {
  fn from_ref(value: impl AsRef<str>) -> Self {
    Self(value.as_ref().to_string())
  }
  fn get_skribble_value(&self) -> &String {
    &self.0
  }
}

pub trait MediaQueryMotion: SkribbleValue {
  /// The media query for devices with a reduced motion preference.
  fn motion_reduce(&self) -> MediaQueryMotionChild {
    MediaQueryMotionChild::from_ref(self.append_to_skribble_value("motionReduce"))
  }
  /// The media query for devices with a no preference motion preference.
  fn motion_safe(&self) -> MediaQueryMotionChild {
    MediaQueryMotionChild::from_ref(self.append_to_skribble_value("motionSafe"))
  }
}
pub struct MediaQueryDarkModeChild(String);
impl SkribbleValue for MediaQueryDarkModeChild {
  fn from_ref(value: impl AsRef<str>) -> Self {
    Self(value.as_ref().to_string())
  }
  fn get_skribble_value(&self) -> &String {
    &self.0
  }
}

pub trait MediaQueryDarkMode: SkribbleValue {
  /// The media query for devices with a dark color scheme.
  fn dark(&self) -> MediaQueryDarkModeChild {
    MediaQueryDarkModeChild::from_ref(self.append_to_skribble_value("dark"))
  }
  /// The media query for devices with a light color scheme.
  fn light(&self) -> MediaQueryDarkModeChild {
    MediaQueryDarkModeChild::from_ref(self.append_to_skribble_value("light"))
  }
}
impl MediaQueryDeviceCategories for SkribbleRoot {}
impl MediaQueryBreakpoints for SkribbleRoot {}
impl MediaQueryOrientation for SkribbleRoot {}
impl MediaQueryMotion for SkribbleRoot {}
impl MediaQueryDarkMode for SkribbleRoot {}
impl MediaQueryBreakpoints for MediaQueryDeviceCategoriesChild {}
impl MediaQueryOrientation for MediaQueryDeviceCategoriesChild {}
impl MediaQueryMotion for MediaQueryDeviceCategoriesChild {}
impl MediaQueryDarkMode for MediaQueryDeviceCategoriesChild {}
impl MediaQueryOrientation for MediaQueryBreakpointsChild {}
impl MediaQueryMotion for MediaQueryBreakpointsChild {}
impl MediaQueryDarkMode for MediaQueryBreakpointsChild {}
impl MediaQueryMotion for MediaQueryOrientationChild {}
impl MediaQueryDarkMode for MediaQueryOrientationChild {}
impl MediaQueryDarkMode for MediaQueryMotionChild {}
